### DISCUSSION TITLE
Calculating area of rasterized square

### DISCUSSION THREAD

**Discussion by trygvels** [corteva/geocube#120] [Q&A] [answered]
Hi!
First of all, I very much appreciate your work, it is great!

My question is, is it possible to calculate the area of the polygon that one is gridding using the make_geocube method and save it to the data-column?

Once I convert my geometry-column I lose all information about the area of the layer, but I would like to know what percentage of a particular square was filled with my geometry layer.

Cheers,
Trygve

**Comment by snowman2** [comment, marked as answer]
It is likely not easy to get. One method would be to polygonize the raster cells and then intersect them with the input geometry. The output of that could be used to calculate the overlap percentage.

**Reply by trygvels** [reply]
Fantastic idea, Alan! Haha, and that's exactly what I did.
The only issue is I have 400 million squares, so the overlap takes a long while. I'm not sure how to speed it up more than I do already.

After polygonizing the raster data to a grid, I get df_grid and do the following
```Python
# First I overlay the polygonized grid on my polygon-data-layer.
# This is the most time consuming part
_overlay = gpd.overlay(layer, df_grid, how="intersection").reset_index()

# Once I have split the layer into finer chunks that fit into each square, I calculate the area
_overlay["area"] = _overlay.area

# Then I join each layer chunk back into their respective squares
joined = gpd.sjoin(df_grid, _overlay, how="left")
# Make sure the geometry is valid
joined["geometry"] = joined["geometry"].buffer(0)

# I finally sum the area within each square
scored_grid = (
      joined[["geometry", "area",]].dissolve(joined.index, sort=False).set_crs(25833)
  )
```

This might not by up your alley, but maybe you have an idea, or perhaps know where I could ask? :)
Could I perhaps paralelnize it?
Thanks again! Geocube already sped up my gridding by a factor of a hundred...

**Reply by snowman2** [reply]
What you have makes sense to me. The only thoughts I have are to try shapely 2.0 or pygeos with geopandas and/or try https://github.com/geopandas/dask-geopandas.

**Reply by trygvels** [reply]
I managed to get a few more speedups with the following code (Posting here for anyone who might have a similar problem):
```Python
def dissolve_shuffle(ddf, by=None, **kwargs):
    """Shuffle and map partition"""
    meta = ddf._meta.dissolve(by=by, as_index=False, **kwargs)
    shuffled = ddf.shuffle(
        by, npartitions=ddf.npartitions, shuffle="tasks", ignore_index=True
    )
    return shuffled.map_partitions(
        gpd.GeoDataFrame.dissolve, by=by, as_index=False, meta=meta, **kwargs
    )

def calculate_area_overlap_dask(area_name, df_grid, layer, nthreads=8,):
    # Reduce dataframe
    layer = layer[["geometry"]]
    df_grid = df_grid[["geometry"]]

    # Overlay layer on grid
    _overlay = gpd.overlay(layer, df_grid, how="intersection")
    _overlay = dg.from_geopandas(_overlay, npartitions=nthreads)

    # calculate area of split layer
    _overlay["area"] = _overlay.area
    _overlay = _overlay.compute()
    df_grid = dg.from_geopandas(df_grid, npartitions=nthreads)

    # assign smaller chunks to respective squares
    joined = dg.sjoin(df_grid, _overlay, how="inner").reset_index()

    # sum polygons in grid using smart suffle
    scored_grid = dissolve_shuffle(joined, "index",)
    scored_grid = scored_grid.compute()
    scored_grid = scored_grid.set_crs(25833)
    return scored_grid
```


This sped it up quite a bit. I am currently looking into your suggestions, but does shapely 2.0 support GeoPandas? I get a bunch of errors if a upgrade it.

**Reply by snowman2** [reply]
Possible that geopandas hasn't made the updates to be compatible yet. I see this PR: https://github.com/geopandas/geopandas/pull/2570

**Reply by trygvels** [reply]
Hmm, it seems different to the issue I'm encountering. 
I get `AttributeError: module 'shapely.geos' has no attribute 'lgeos'`, so there seems to be some in PYGEOS_SHAPELY_COMPAT, which is called when I read a file with Fiona as the engine.

**Reply by snowman2** [reply]
This may be the PR: https://github.com/geopandas/geopandas/pull/2275
